aip/general/0193.md

```diff
@@ -75,132 +86,191 @@ following:
+##### [google.rpc.Status payload][Status]
 
-For each error, the service **must** populate the `message` field on
-[`google.rpc.Status`][Status]. This error message,
+Details of this object are summarized in the following fields, field
+descriptions, and examples:
 
-- is a developer-facing, human-readable "debug message"
-- is presented in the service's native language
-- both explains the error and offers an actionable resolution to it
-  ([citation](https://cloud.google.com/apis/design/errors#error_model))
 
-**Note:** Sometimes a service will elect to always present
-`Status.message` in English rather than the application's native
-language so that messages are easily searchable in common knowledge
-bases, such as StackOverflow&trade;.
+- `code int32`
+  - The status code, which **should** be an enum value of
+    [`google.rpc.Code`][Code].
+  - *Example*: 429
+- `message string`
+  - A developer-facing error message, which **should** be in English.
+    Any user-facing error message **should** be located in the
+    LocalizedMessage payload.
+  - *Example message*: \"The zone \'us-east1-a\' does not have enough
+    resources available to fulfill the request. Try a different zone, or
+    try again later.\"
+- `details Object (repeated)`
+  - Text
+  - *Example*: Described in the [ErrorDetails](#error-details) section.
 
-When introducing an error that represents a failure scenario that did
-not previously occur for the service, the payload **must** include
-`ErrorInfo` and any variables found in dynamic segments of the error
-message **must** be present in `ErrorInfo.metadata`. See, "[Dynamic
-variables](#dynamic-variables)".
 
-#### Changing error messages
+#### [ErrorDetails][details]
 
-<a name="status-message-warning"></a>
+Google defines a set of [standard detail payloads][details] for error
+details, which cover most common needs for API errors.  Services
+**should** use these standard detail payloads when feasible.
 
-`Status.message` **may** change. However, **use extreme caution** when
-doing so.  API consumers often treat this error message as **part of the
-API contract**. Clients perform string matches on the text to
-differentiate one error for another and even parse the error message to
-extract variables from dynamic segments.
+Structured details with machine-readable identifiers **must** be used if
+users will need to write code against a specific aspect of the error.
+Error message strings **may** change over time; however, if an error
+message does not have a machine-readable identifier *in addition to* the
+`code` string, changing the error message **must** be considered a
+backwards-incompatible change.
 
-There is a safer alternative. The service can chose to include an error
-message by adding [`google.rpc.LocalizedMessage`][LocalizedMessage] to
-[`Status.details`][Status details].
+For the purpose of this discussion, the ErrorDetails section contains
+three payloads: `ErrorInfo`, `LocalizedMessage`, and `HelpLink`.
```

Comment:
This is not useful or true. It can contain anything and often does have BadRequest or PreconditionFailure, etc.

Reply:
@shwoodard @sgiovanni1 would it make sense to change this to say:

> ErrorDetails **must** contain one instance of `ErrorInfo`, one `LocalizedMessage`, and one `HelpLink`, and **may** contain one instance of additional error detail types as defined by [google/rpc/error_details.proto.](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto).

Please correct me if I'm wrong on whether `LocalizedMessage` and `HelpLink` are required, I'm pretty sure `ErrorInfo` is.

Reply:
@andrei-scripniciuc It seems to me that this sentence is merely an introduction to the following sections that describes `ErrorInfo`, `LocalizedMessage`, and `HelpLink`. Do we need to spell out what must be included? A different way to position this is something like below.

In the following section, we will describe `ErrorInfo`, `LocalizedMessage`, and `HelpLink` in more detail.

Reply:
Updated. "The following section describes `ErrorInfo`, `LocalizedMessage`, and `HelpLink`
in more detail."